pkg/apis/ibmcloudprovider/v1beta1/ibmcloudproviderconfig_types.go

```diff
@@ -0,0 +1,107 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IBMCloudMachineProviderSpec is the type that will be embedded in a Machine.Spec.ProviderSpec field
+// for an IBM Cloud VPC virtual machine. It is used by the IBM Cloud machine actuator to create a single Machine.
+// +k8s:openapi-gen=true
+type IBMCloudMachineProviderSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// vpc type Instance struct
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The user-defined name for this virtual server instance
+	// default: system hostname
+	Name string `json:"name,omitempty"`
+
+	// VPC name where the instance will be created
+	VPC string `json:"vpc"`
+
+	// Actuator will apply these tags to an virtual server instance if not present in additon
+	// to default tags applied by the actuator
+	// TODO
+	// Tags []string `json:"tags,omitempty"`
+
+	// TODO: Add labels to the virtual server
+	// Labels map[string]string `json:"labels,omitempty"`
+
+	// Image is the id of the custom OS image in VPC
+	// Example: a123-ade15224-ed7y-5e44-ye56-8199d4557699
+	Image string `json:"image"`
+
+	// Profile indicates the flavor of instance.
+	// Example: bx2-8x32 (8 vCPUs, 32 GB RAM)
+	Profile string `json:"profile"`
+
+	// Zone where the virtual server instance will be created
+	Zone string `json:"zone"`
+
+	// ProviderID is the unique identifier as specified by the cloud provider.
+	// +optional
+	ProviderID string `json:"providerID,omitempty"`
+
+	// PrimaryNetworkInterface is required to specify subnet
+	PrimaryNetworkInterface NetworkInterface `json:"primaryNetworkInterface,omitempty"`
+
+	// TODO: Probably not needed for the worker machines
+	// SSHKeys is the SSH pub keys that will be used to access virtual service instance
+	// SSHKeys []*string `json:"sshKeys,omitempty"`
+
+	// UserDataSecret holds reference to a secret which containes Instance Ignition data (User Data)
+	UserDataSecret *corev1.LocalObjectReference `json:"userDataSecret,omitempty"`
+
+	// CredentialsSecret is a reference to the secret with IBM Cloud credentials.
+	CredentialsSecret *corev1.LocalObjectReference `json:"credentialsSecret,omitempty"`
+}
+
+// NetworkInterface struct
+type NetworkInterface struct {
+	// Subnet ID of the network interface
+	Subnet string `json:"subnet,omitempty"`
```

Comment:
@gsrajpal Should the Subnet attribute here be of type Subnet defined below? Rather than a string?

Reply:
Good catch -- Don't think we need that `type Subnet` below actually.

We only need the Subnet ID so this should suffice.

Reply:
@gsrajpal Thinking about this a little more, should this accept Subnet names as well? From the installer's perspective, it generates all the manifests (including MachineSets and Machines) before Terraform is executed. In this case, we would not know the Subnet ID beforehand.

Reply:
Yeah that sounds right, Hide for the same reason we need name in the Image field.

Reply:
We could work w/ just the name, if that's all we have. Just need to lookup the id, href, or crn of the subnet when provisioning the vsi. I wonder if there's some post-terraform hook we could leverage to populate some of these IDs after the fact. Names aren't great, since they can be changed later on.

Reply:
As far as I am aware, there is not such a thing.

Reply:
The installer will be able to provide the subnet name, VPC name, region, zone, and resource group name.

Reply:
We can write a logic to grab the id if you want to pass in the subnet name. 

We cannot pass in `Subnet Name` directly to VPC.